### PR TITLE
kubeadm: Group centric component configs

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:

kubeadm's current implementation of component config support is "kind" centric.
This has its downsides. Namely:
- Kind names and numbers can change between config versions.
  Newer kinds can be ignored. Therefore, detection of a version change is
  considerably harder.
- A component config can have only one kind that is managed by kubeadm.

Thus a more appropriate way to identify component configs is required.

Probably the best solution identified so far is a config group.
A group name is unlikely to change between versions, while the kind names and
structure can.
Tracking component configs by group name allows us to:
- Spot more easily config version changes and manage alternate versions.
- Support more than one kind in a config group/version.
- Abstract component configs by hiding their exact structure.

Hence, this change rips off the old kind based support for component configs
and replaces it with a group name based one. This also has the following
extra benefits:
- More tests were added.
- kubeadm now errors out if an unsupported version of a known component group
  is used.

**Which issue(s) this PR fixes**:

Refs:
- kubernetes/kubeadm#1940
- kubernetes/enhancements#1381

**Special notes for your reviewer**:

This PR is part of the implementation of the new kubeadm component config management scheme KEP (see link below).

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-soon
/assign @timothysc @fabriziopandini @neolit123 @ereslibre 

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm now errors out whenever a not supported component config version is supplied for the kubelet and kube-proxy
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md
```

### CODE REVIEW

**Comment by rosti** [comment]
/hold
for discussion.

File: cmd/kubeadm/app/componentconfigs/configset.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+// Validate is a placeholder for performing a validation on an already loaded component configs in a ClusterConfiguration
+// Currently it prints a warning that no validation was performed
+func Validate(clusterCfg *kubeadmapi.ClusterConfiguration) field.ErrorList {
+	groups := []string{}
+	for group := range clusterCfg.ComponentConfigs {
+		groups = append(groups, group)
+	}
+	sort.Strings(groups) // The sort is needed to make the output predictable
+	klog.Warningf("WARNING: kubeadm cannot validate component configs for API groups %v", groups)
```
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/configset.go]: i think we saw this warning being printed in a lot of commands such as `kubeadm config images`.
should we change it to klog.V(...).Infof()?

**Comment by neolit123** [comment]
i'm +1 on this refactor. thanks @rosti 

i envision the following split in the future:
- kubelet, and control-plane component configs being managed by kubeadm in a similar way (as core components)
- kube-proxy moving to addons.

File: cmd/kubeadm/app/componentconfigs/configset.go (hunk 1 above)
Comment (rosti) [reply cmd/kubeadm/app/componentconfigs/configset.go]: I guess so. The code is executed whenever the config is loaded (which happens in almost all kubeadm commands). However it's good to see it as a warning only on `kubeadm init/upgrade *`.

**Comment by rosti** [comment]
> i'm +1 on this refactor. thanks @rosti
> 
> i envision the following split in the future:
> 
> * kubelet, and control-plane component configs being managed by kubeadm in a similar way (as core components)
> * kube-proxy moving to addons.

Exactly, that's my vision too.

**Comment by ereslibre** [comment]
On a first sight this SGTM. I plan to review this more thoroughly later today though.

File: cmd/kubeadm/app/componentconfigs/configset.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return h.FromDocumentMap(gvkmap)
+}
+
+// FromCluster loads a component from a config map in the cluster
+func (h *handler) FromCluster(clientset clientset.Interface, clusterCfg *kubeadmapi.ClusterConfiguration) (kubeadmapi.ComponentConfig, error) {
+	return h.fromCluster(h, clientset, clusterCfg)
```
Comment (ereslibre) [inline cmd/kubeadm/app/componentconfigs/configset.go]: Can't we make `fromCluster` a `handler` method as well, so there's no need to pass `h` as argument?

**Review by ereslibre — COMMENTED** [review]
LGTM, is a big chunk to review, but I'm +1 overall. A small comment for a tiny improvement.

File: cmd/kubeadm/app/componentconfigs/configset.go (hunk 2 above)
Comment (ereslibre) [reply cmd/kubeadm/app/componentconfigs/configset.go]: I see why you did this on the non-public one... nevermind.

**Comment by ereslibre** [comment]
/lgtm
/hold
^ for other reviewers.

**Review by yastij — COMMENTED** [review]
/lgtm

**Comment by neolit123** [comment]
/hold cancel
this PR received 3 +1s.

**Pull request merged**